src/controller/testController.ts

```diff
@@ -265,23 +265,90 @@ function enqueueTestMethods(testItems: TestItem[], run: TestRun): void {
-   // @ts-expect-error
-   const excludingItems: TestItem[] = await expandTests(request.exclude || [], TestLevel.Class, token);
-   testItems = _.differenceBy(testItems, excludingItems, 'id');
-   return testItems;
+    let testItems: TestItem[] = [];
+    if (request.include) {
+        testItems.push(...request.include);
+    } else {
+        testController?.items.forEach((item: TestItem) => {
+            testItems.push(item);
+        });
+    }
+    if (testItems.length === 0) {
+        return [];
+    }
+    testItems = handleInvocations(testItems);
+    testItems = await expandTests(testItems, TestLevel.Class, token);
+    // @ts-expect-error
+    const excludingItems: TestItem[] = await expandTests(request.exclude || [], TestLevel.Class, token);
+    testItems = _.differenceBy(testItems, excludingItems, 'id');
+    return testItems;
+}
+
+/**
+ * Check and preparation in case a single invocation of a parameterized test is re-run.
+ * If a test is run completely, existing invocations are removed.
+ * @param testItems
+ * @returns prepared testItems
+ */
+function handleInvocations(testItems: TestItem[]): TestItem[] {
+
+    // always remove uniqueIds from non-invocation items, since they would have been set for a past run
+    testItems.forEach((item: TestItem) => {
+        const itemData: ITestItemData | undefined = dataCache.get(item);
+        if (itemData && itemData.testLevel !== TestLevel.Invocation) {
+            itemData.uniqueId = undefined;
+        }
+    });
+
+    if (filterInvocations(testItems).some((invocation: TestItem)
+        => !invocation.parent || !dataCache.get(invocation.parent))) { // sanity-checks
+        const errMsg: string = 'Trying to re-run a single test invocation, but could not find a corresponding method-level parent item with data.';
+        sendError(new Error(errMsg));
+        window.showErrorMessage(errMsg);
+        return [];
+    }
+
+    // remove invocations if they are already included in selected higher-level tests
+    testItems = testItems.filter((item: TestItem) => !(isInvocation(item) && isAncestorIncluded(item, testItems)));
+
+    const invocations: TestItem[] = filterInvocations(testItems);
+    if (invocations.length > _.uniq(invocations.map((item: TestItem) => item.parent)).length) {
+        window.showErrorMessage('Re-running multiple invocations of a parameterized test is not supported, please select only one invocation at a time.');
+        return [];
+    }
+
+    // if a single invocation is to be re-run,
+    // we run the parent method instead, but with restriction to the single invocation parameter-set
+    testItems = testItems.map((item: TestItem) => {
+        if (isInvocation(item)) {
+            dataCache.get(item.parent!)!.uniqueId = dataCache.get(item)!.uniqueId;
+            return item.parent;
+        }
+        return item;
+    })
+
+    removeNonRerunTestInvocations(testItems);
+    return testItems;
+}
```

Comment:
> Another issue I found is after running a parameterized tests, return it from the method level will throw an error.
This is because VS Code will give us the method item and all its invocation nodes. Seems that we need to try to merge the test items before the statement const numInvocations: number = testItems.filter((item: TestItem) => dataCache.get(item)?.testLevel === TestLevel.Invocation).length;.

@jdneo The selected invocations are now removed normally in case they are included in another selected higher-level item (e.g. when run from method code lens):
![run_all_from_method_code_lens](https://user-images.githubusercontent.com/53265832/164225182-460348ed-0909-4beb-9148-b447947a25b0.gif)

Also it is now possible to re-run invocations when they have different parents:
![rerun_with_different_parents](https://user-images.githubusercontent.com/53265832/164225344-e2b9b3c4-afd8-4c22-9b9c-a2d6dcc1fe7e.gif)
 
Mixing invocation re-run and method run is also handled:
![rerun_with_and_without_parent](https://user-images.githubusercontent.com/53265832/164225972-327efc93-2b25-4f0c-9471-ec7dcb789d82.gif)

Re-running two invocations from the same method will cause the error:
![rerun_multiple_error](https://user-images.githubusercontent.com/53265832/164226212-ff46fa39-29b1-46f2-a1a8-70a7aa6d10f3.gif)